### PR TITLE
[DALI TF] Try building from source when TF version doesn't match exactly. Add test step to installation script.

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

It enhances the DALI TF plugin installation logic in two aspects:
- When no exact TF version match, always try to build from source before trying to install a prebuilt plugin
- Add a test step at the end of the installation, verifying that the plugin works. Test step is skipped if DALI not yet installed.

### Affected modules and functionalities:

DALI TF installation script

### Key points relevant for the review:

Installation logic

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2606

### CODE REVIEW

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 1]
```diff
@@ -243,6 +223,78 @@ def install(self):
...
+        test_batch = get_data()
+        with Session() as sess:
+            print(sess.run(test_batch))
+
+    def install(self):
+        self.check_install_env()
+
+        if self.prebuilt_exact_ver:
+            self.install_prebuilt()
+        else:
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:281]:
```suggestion
        if self.prebuilt_exact_ver:
            self.install_prebuilt()
            try:
                self.test()
                return
            except ImportError as e:
                pass
```

Comment (JanuszL) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py:281]: Maybe we should test prebuild as well, and if it fails try to build from source.

**Comment by jantonguirao** [comment]
!build

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 2]
```diff
@@ -243,6 +223,86 @@ def install(self):
...
+                print("Error while testing the prebuilt plugin: ", e)
+                # Will try building from source
+
+        try:
+            self.build()
+        except Exception as e:
+            print("Build from source failed with error: ", e)
+            if self.can_install_prebuilt:
+                print("Trying to install prebuilt plugin")
+                self.install_prebuilt()
```
Comment (stiepan) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:300]: So if `prebuilt_exact_ver` and `can_install_prebuilt` are True, we may have three attempts at installing DALI: `install_prebuilt`, `build` and again `install_prebuilt`?

Comment (jantonguirao) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py]: True. Perhaps we should only try the second install prebuilt if there was no first attempt.

Comment (jantonguirao) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py]: Done

**Review by JanuszL — APPROVED** [review]

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 3]
```diff
@@ -239,10 +287,40 @@ def install(self):
...
-                raise ImportError("Error while importing the DALI TF plugin built from source, will not install")
+            if not self.check_plugin(lib_path, dali_stub_lib):
+                raise ImportError("Error while loading or testing the DALI TF plugin built from source, will not install")
             print("Installation successful")
 
+    def check_install_env(self):
+        print("Checking build environment for DALI TF plugin ...")
+        print(self.debug_str())
+
+        Path(self.plugin_dest_dir).mkdir(parents=True, exist_ok=True)
```
Comment (stiepan) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:298]: nitpick: Maybe it's better to first make a check that may raise an error and only if it passes, create the directory?

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 4]
```diff
@@ -239,10 +287,40 @@ def install(self):
...
+            error_msg += "\n Tensorflow installation not found. Install `tensorflow-gpu` and try again"
+            error_msg += '\n' + self.debug_str()
+            raise ImportError(error_msg)
+
+    def install(self):
+        self.check_install_env()
+
+        if self.prebuilt_exact_ver:
+            if self.install_prebuilt():
+                return
```
Comment (stiepan) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:311]: Probably very nitpicky

```suggestion
        if self.prebuilt_exact_ver and self.install_prebuilt():
           return
```

**Review by stiepan — APPROVED** [review]

**Pull request merged**